File: services/auth-service/src/monitoring/prometheus.ts
import client from 'prom-client';
import { Request, Response, NextFunction } from 'express';

const collectDefaultMetrics = client.collectDefaultMetrics;
collectDefaultMetrics({ register: client.register });

export const httpRequestCounter = new client.Counter({
  name: 'http_requests_total',
  help: 'Total number of HTTP requests',
  labelNames: ['method', 'route', 'status_code']
});

export const metricsMiddleware = (req: Request, res: Response, next: NextFunction) => {
  res.on('finish', () => {
    httpRequestCounter.labels(
      req.method, 
      req.route ? req.route.path : req.path, 
      res.statusCode.toString()
    ).inc();
  });
  next();
};

export const metricsHandler = async (req: Request, res: Response) => {
  res.set('Content-Type', client.register.contentType);
  res.end(await client.register.metrics());
};

File: services/workspace-service/pkg/monitoring/prometheus.go
package monitoring

import (
	"strconv"

	"github.com/gin-gonic/gin"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
	"github.com/prometheus/client_golang/prometheus/promhttp"
)

var HTTPRequestCounter = promauto.NewCounterVec(
	prometheus.CounterOpts{
		Name: "http_requests_total",
		Help: "Total number of HTTP requests",
	},
	[]string{"method", "route", "status_code"},
)

func MetricsHandler() gin.HandlerFunc {
	h := promhttp.Handler()
	return func(c *gin.Context) {
		h.ServeHTTP(c.Writer, c.Request)
	}
}

func MetricsMiddleware() gin.HandlerFunc {
	return func(c *gin.Context) {
		c.Next()

		routePattern := c.FullPath()
		if routePattern == "" {
			routePattern = c.Request.URL.Path
		}

		statusCode := strconv.Itoa(c.Writer.Status())

		HTTPRequestCounter.WithLabelValues(
			c.Request.Method,
			routePattern,
			statusCode,
		).Inc()
	}
}
